### PR TITLE
CORE & PROV: make export of key data leaner through callback, take 2

### PR DESCRIPTION
**Pull request by levitte**
Exporting data from a provider owned domainparams or key is quite an
ordeal, with having to figure out what parameter keys an
implementation supports, call the export function a first time to find
out how large each parameter buffer must be, allocate the necessary
space for it, and call the export function again.

So how about letting the export function build up the key data params
and call back with that?  This change implements exactly such a
mechanism.

-----

This is an alternative to #10391

### CODE REVIEW

**Comment by mattcaswell** [comment]
Travis failures are relevant

**Comment by levitte** [comment]
Darnit

**Comment by levitte** [comment]
Oh wait, that's because #10412 isn't merged yet, and I haven't included it in this PR.
Comments on the code and the idea would still be appreciated.

File: crypto/evp/keymgmt_lib.c
Comment (mattcaswell) [inline crypto/evp/keymgmt_lib.c]: With this deletion we never seem to call the `exportkey_types()` keymgmt function anywhere. Similarly we never seem to use the other `[export|import][key|domparams]_types()` functions. Should they be deleted?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
This looks a lot neater. I just have one comment.

File: crypto/evp/keymgmt_lib.c
Comment (levitte) [reply crypto/evp/keymgmt_lib.c]: Can we leave that undecided for now? I'm thinking of provider exploration with `openssl provider`, but you may be right.

It will remain on my mind, so post-pone to a cleanup PR to give me time to think? Feel free to hold me to it :wink:

Comment (mattcaswell) [reply crypto/evp/keymgmt_lib.c]: Could we add a `TODO` in the code as a placeholder so that we do remember to come back and consider it at some point?

Comment (levitte) [reply crypto/evp/keymgmt_lib.c]: Yes

Comment (levitte) [reply crypto/evp/keymgmt_lib.c]: Done

**Review by mattcaswell — APPROVED** [review]
Approved subject to a clean CI run once #10412 is merged.

**Comment by levitte** [comment]
#10412 now being merged, I've rebased this

**Comment by levitte** [comment]
CIs are happy, apart from the Appveyor things that aren't relevant here

**Comment by levitte** [comment]
Merged.

1640d48c5b4ee0a3ff5a2a5015ee17ac163d9cd4 CORE & PROV: make export of key data leaner through callback

**Pull request closed without merging**
